### PR TITLE
Initialise comments

### PR DESCRIPTION
**Pull request by rishabhsaxena**
## Description
Initialise comments:
- Config flags in place to keep the feature hidden / not in the execution path
- Overlay comments wrapper at the base-widget, that renders comments using `position: absolute`
- Unpublished comment that is pushed to the server once the comment is actually added, until then it's just within the store 
- Normalised store structure to prevent re-renders across all widgets when a comment is added
- Overlay over components except a blacklist to prevent interaction with the components while adding comments, the blacklist is added for container components
- Toggle comment mode button at the header toggles the comment mode
- Websockets - a socket connection is initiated after authentication / disconnected on logout, we maintain socket connection status in the store, and an action to reconnect is supported, but not accessible from the UI atm

An additional change dispatches logoutAction on `401`, to reset store and disconnect socket.

## Type of change
- New feature (non-breaking change which adds functionality)

## How Has This Been Tested?
- Integration tests for`OverlayCommentsWrapper` with API mocks for:
  - Rendering threads
  - Creating a new thread
  - Replying to a thread 

## Checklist:

- [x] My code follows the style guidelines of this project
- [x] I have performed a self-review of my own code
- [x] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation
- [x] My changes generate no new warnings
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] New and existing unit tests pass locally with my changes

## Test coverage results :test_tube:
<details><summary>:green_circle: Total coverage has increased</summary>

    // Code coverage diff between base branch:release and head branch: init-comments 
Status | File | % Stmts | % Branch | % Funcs | % Lines 
 -----|-----|---------|----------|---------|------ 
 :green_circle: | total | 50.09 **(1.39)** | 30.62 **(0.95)** | 29.02 **(1.75)** | 50.49 **(1.33)**
 :sparkles: :new: | **app/client/src/actions/commentActions.ts** | **87.84** | **100** | **62.5** | **89.47**
 :green_circle: | app/client/src/actions/initActions.ts | 80 **(10)** | 100 **(0)** | 33.33 **(33.33)** | 80 **(0)**
 :sparkles: :new: | **app/client/src/actions/websocketActions.ts** | **71.43** | **100** | **0** | **75**
 :green_circle: | app/client/src/api/Api.ts | 83.33 **(8.33)** | 25 **(12.5)** | 60 **(20)** | 83.33 **(8.33)**
 :green_circle: | app/client/src/api/ApiUtils.ts | 65.38 **(9.61)** | 45.71 **(11.42)** | 80 **(0)** | 62.5 **(10.42)**
 :sparkles: :new: | **app/client/src/api/CommentsAPI.tsx** | **80.95** | **100** | **57.14** | **80**
 :green_circle: | app/client/src/api/OrgApi.ts | 59.38 **(3.13)** | 0 **(0)** | 16.67 **(8.34)** | 58.06 **(3.22)**
 :sparkles: :new: | **app/client/src/comments/ToggleCommentModeButton.tsx** | **65.85** | **37.5** | **40** | **65.85**
 :sparkles: :new: | **app/client/src/comments/utils.ts** | **84.62** | **0** | **50** | **77.78**
 :sparkles: :new: | **app/client/src/comments/CommentCard/CommentCard.tsx** | **86.36** | **77.78** | **84.21** | **86.15**
 :sparkles: :new: | **app/client/src/comments/CommentCard/CommentContextMenu.tsx** | **75** | **100** | **42.86** | **71.88**
 :sparkles: :new: | **app/client/src/comments/CommentCard/ResolveCommentButton.tsx** | **100** | **83.33** | **100** | **100**
 :sparkles: :new: | **app/client/src/comments/CommentThread/CommentThread.tsx** | **94.74** | **78.38** | **90** | **94.74**
 :sparkles: :new: | **app/client/src/comments/CommentThread/ScrollToLatest.tsx** | **100** | **100** | **100** | **100**
 :sparkles: :new: | **app/client/src/comments/inlineComments/AddCommentInput.tsx** | **79.73** | **85.71** | **68.75** | **79.03**
 :sparkles: :new: | **app/client/src/comments/inlineComments/Comments.tsx** | **100** | **100** | **100** | **100**
 :sparkles: :new: | **app/client/src/comments/inlineComments/InlineCommentPin.tsx** | **87.18** | **33.33** | **88.89** | **86.11**
 :sparkles: :new: | **app/client/src/comments/inlineComments/OverlayCommentsWrapper.tsx** | **96.97** | **75** | **100** | **100**
 :sparkles: :new: | **app/client/src/comments/inlineComments/StyledComponents.tsx** | **100** | **100** | **100** | **100**
 :sparkles: :new: | **app/client/src/comments/inlineComments/UnpublishedCommentThread.tsx** | **89.66** | **50** | **71.43** | **88.89**
 :sparkles: :new: | **app/client/src/comments/inlineComments/useOrgUsers.tsx** | **100** | **100** | **100** | **100**
 :sparkles: :new: | **app/client/src/comments/tests/setMockPages.ts** | **100** | **100** | **100** | **100**
 :sparkles: :new: | **app/client/src/components/ads/EmojiPicker.tsx** | **82.35** | **100** | **33.33** | **80**
 :green_circle: | app/client/src/components/ads/Icon.tsx | 60.79 **(3.29)** | 42.11 **(3.29)** | 100 **(0)** | 60.62 **(3.33)**
 :sparkles: :new: | **app/client/src/components/ads/MentionsInput.tsx** | **87.5** | **58.33** | **71.43** | **85.29**
 :green_circle: | app/client/src/components/ads/Toast.tsx | 45.65 **(10.87)** | 34.62 **(19.24)** | 5.88 **(5.88)** | 45.65 **(10.87)**
 :sparkles: :new: | **app/client/src/components/editorComponents/PreventInteractionsOverlay.tsx** | **100** | **87.5** | **100** | **100**
 :sparkles: :new: | **app/client/src/constants/CommentConstants.tsx** | **100** | **100** | **100** | **100**
 :green_circle: | app/client/src/constants/DefaultTheme.tsx | 93 **(0.14)** | 88.89 **(0)** | 75 **(0)** | 91.67 **(0.21)**
 :sparkles: :new: | **app/client/src/constants/WebsocketConstants.tsx** | **70** | **100** | **0** | **100**
 :green_circle: | app/client/src/constants/messages.ts | 70.64 **(0.6)** | 100 **(0)** | 11.11 **(1.86)** | 74.03 **(0.72)**
 :sparkles: :new: | **app/client/src/mockResponses/CommentApiMockResponse.tsx** | **100** | **100** | **100** | **100**
 :red_circle: | app/client/src/pages/Applications/ForkApplicationModal.tsx | 87.27 **(-7.28)** | 76 **(-20)** | 50 **(-20)** | 86.79 **(-7.55)**
 :green_circle: | app/client/src/pages/Editor/EditorHeader.tsx | 87.21 **(0.15)** | 46.34 **(0)** | 73.91 **(0)** | 86.9 **(0.15)**
 :green_circle: | app/client/src/reducers/entityReducers/pageListReducer.tsx | 19.44 **(5.55)** | 27.78 **(27.78)** | 15.38 **(15.38)** | 21.88 **(6.25)**
 :green_circle: | app/client/src/reducers/uiReducers/editorReducer.tsx | 26.87 **(2.99)** | 0 **(0)** | 13.64 **(9.09)** | 26.87 **(2.99)**
 :green_circle: | app/client/src/reducers/uiReducers/errorReducer.tsx | 77.78 **(11.11)** | 100 **(0)** | 33.33 **(33.33)** | 77.78 **(11.11)**
 :green_circle: | app/client/src/reducers/uiReducers/orgReducer.ts | 16.22 **(2.71)** | 0 **(0)** | 4.17 **(4.17)** | 16.22 **(2.71)**
 :sparkles: :new: | **app/client/src/reducers/uiReducers/websocketReducer.ts** | **83.33** | **100** | **0** | **83.33**
 :sparkles: :new: | **app/client/src/reducers/uiReducers/commentsReducer/commentsReducer.ts** | **67.39** | **50** | **76.47** | **75.61**
 :sparkles: :new: | **app/client/src/reducers/uiReducers/commentsReducer/handleAddCommentToThreadSuccess.ts** | **100** | **100** | **100** | **100**
 :sparkles: :new: | **app/client/src/reducers/uiReducers/commentsReducer/handleCreateNewCommentThreadSuccess.ts** | **92.31** | **50** | **100** | **90**
 :sparkles: :new: | **app/client/src/reducers/uiReducers/commentsReducer/handleFetchApplicationCommentsSuccess.ts** | **93.33** | **75** | **100** | **100**
 :sparkles: :new: | **app/client/src/reducers/uiReducers/commentsReducer/handleNewCommentThreadEvent.ts** | **92.31** | **75** | **100** | **92.31**
 :sparkles: :new: | **app/client/src/reducers/uiReducers/commentsReducer/handleUpdateCommentThreadEvent.ts** | **100** | **100** | **100** | **100**
 :sparkles: :new: | **app/client/src/reducers/uiReducers/commentsReducer/handleUpdateCommentThreadSuccess.ts** | **100** | **100** | **100** | **100**
 :sparkles: :new: | **app/client/src/reducers/uiReducers/commentsReducer/testFixtures.ts** | **100** | **100** | **100** | **100**
 :green_circle: | app/client/src/sagas/ApplicationSagas.tsx | 24.26 **(0.31)** | 7.84 **(0)** | 15.79 **(0)** | 27.08 **(0.32)**
 :green_circle: | app/client/src/sagas/ErrorSagas.tsx | 65.09 **(25.47)** | 42.11 **(24.57)** | 50 **(27.78)** | 65.26 **(26.31)**
 :green_circle: | app/client/src/sagas/InitSagas.ts | 24.04 **(0.28)** | 3.33 **(0)** | 20 **(0)** | 28.09 **(0.18)**
 :green_circle: | app/client/src/sagas/OrgSagas.ts | 22.05 **(4.73)** | 3.85 **(0)** | 14.29 **(7.15)** | 25.74 **(3.96)**
 :sparkles: :new: | **app/client/src/sagas/WebsocketSagas.ts** | **19.4** | **9.52** | **13.33** | **23.64**
 :green_circle: | app/client/src/sagas/index.tsx | 90.7 **(0.46)** | 50 **(0)** | 100 **(0)** | 92.86 **(0.36)**
 :red_circle: | app/client/src/sagas/userSagas.tsx | 17.37 **(-0.1)** | 2.33 **(-0.37)** | 8.33 **(0)** | 20.57 **(-0.14)**
 :sparkles: :new: | **app/client/src/sagas/CommentSagas/handleCommentEvents.ts** | **25** | **0** | **0** | **25**
 :sparkles: :new: | **app/client/src/sagas/CommentSagas/index.ts** | **63.75** | **55** | **66.67** | **64.52**
 :green_circle: | app/client/src/selectors/appViewSelectors.tsx | 68.75 **(12.5)** | 0 **(0)** | 28.57 **(28.57)** | 66.67 **(6.67)**
 :sparkles: :new: | **app/client/src/selectors/commentsSelectors.ts** | **59.62** | **25** | **53.33** | **57.58**
 :green_circle: | app/client/src/selectors/editorSelectors.tsx | 59.23 **(0.77)** | 27.03 **(0)** | 35.71 **(2.38)** | 57.01 **(0.94)**
 :green_circle: | app/client/src/selectors/organizationSelectors.tsx | 67.44 **(2.32)** | 0 **(0)** | 26.32 **(5.27)** | 62.96 **(0)**
 :sparkles: :new: | **app/client/src/utils/hooks/useIsScrolledToBottom.tsx** | **100** | **72.73** | **100** | **100**
 :green_circle: | app/client/src/widgets/BaseWidget.tsx | 79.65 **(0.04)** | 39.02 **(0)** | 78.57 **(1.65)** | 78.7 **(0.13)**
 :green_circle: | app/client/test/customQueries.js | 95 **(5)** | 100 **(0)** | 66.67 **(33.34)** | 94.74 **(5.27)**</details>

### CODE REVIEW

**Comment by jrexerpted** [comment]
> ## Test coverage results :test_tube: 
> 
>     Code coverage diff between base branch:release and head branch: init-comments 
> Status | File | % Stmts | % Branch | % Funcs | % Lines 
>  -----|-----|---------|----------|---------|------ 
>  :green_circle: | total | 43.94 **(1.52)** | 27.24 **(0.84)** | 22.36 **(2.2)** | 46.3 **(1.43)**
>  :sparkles: :new: | **app/client/src/actions/commentActions.ts** | **80.95** | **100** | **60** | **96.15**
>  :green_circle: | app/client/src/actions/initActions.ts | 71.43 **(14.29)** | 100 **(0)** | 33.33 **(33.33)** | 80 **(0)**
>  :sparkles: :new: | **app/client/src/actions/websocketActions.ts** | **60** | **100** | **0** | **75**
>  :green_circle: | app/client/src/api/Api.ts | 83.33 **(8.33)** | 25 **(12.5)** | 60 **(20)** | 83.33 **(8.33)**
>  :sparkles: :new: | **app/client/src/api/CommentsAPI.tsx** | **94.44** | **100** | **85.71** | **93.33**
>  :green_circle: | app/client/src/components/ads/Icon.tsx | 56.63 **(4.7)** | 35.9 **(5.76)** | 100 **(0)** | 56.63 **(4.7)**
>  :sparkles: :new: | **app/client/src/components/ads/Comments/AddCommentInput.tsx** | **91.89** | **87.5** | **92.31** | **90.91**
>  :sparkles: :new: | **app/client/src/components/ads/Comments/CommentCard.tsx** | **100** | **100** | **100** | **100**
>  :sparkles: :new: | **app/client/src/components/ads/Comments/Comments.tsx** | **100** | **100** | **100** | **100**
>  :sparkles: :new: | **app/client/src/components/ads/Comments/EmojiPicker.tsx** | **82.35** | **100** | **33.33** | **80**
>  :sparkles: :new: | **app/client/src/components/ads/Comments/InlineCommentPin.tsx** | **100** | **100** | **100** | **100**
>  :sparkles: :new: | **app/client/src/components/ads/Comments/InlineCommentThreadContainer.tsx** | **93.1** | **57.14** | **83.33** | **93.1**
>  :sparkles: :new: | **app/client/src/components/ads/Comments/OverlayCommentsWrapper.tsx** | **93.94** | **62.5** | **100** | **100**
>  :sparkles: :new: | **app/client/src/components/ads/Comments/ResolveCommentButton.tsx** | **100** | **75** | **100** | **100**
>  :sparkles: :new: | **app/client/src/components/ads/Comments/ScrollToLatest.tsx** | **100** | **100** | **100** | **100**
>  :sparkles: :new: | **app/client/src/components/ads/Comments/StyledComponents.tsx** | **100** | **100** | **100** | **100**
>  :sparkles: :new: | **app/client/src/components/ads/Comments/UnpublishedCommentThread.tsx** | **89.66** | **50** | **71.43** | **88.46**
>  :sparkles: :new: | **app/client/src/components/ads/Comments/init.ts** | **45.45** | **100** | **0** | **62.5**
>  :sparkles: :new: | **app/client/src/components/ads/Comments/selectors.ts** | **86.96** | **100** | **72.73** | **90**
>  :sparkles: :new: | **app/client/src/components/ads/Comments/utils.ts** | **80** | **0** | **50** | **77.78**
>  :sparkles: :new: | **app/client/src/components/ads/Comments/tests/setMockPages.ts** | **100** | **100** | **100** | **100**
>  :red_circle: | app/client/src/components/editorComponents/DropTargetComponent.tsx | 15.15 **(-1.18)** | 2.82 **(0.15)** | 0 **(0)** | 16.85 **(-1.33)**
>  :sparkles: :new: | **app/client/src/components/editorComponents/PreventInteractionsOverlay.tsx** | **60** | **50** | **0** | **69.23**
>  :sparkles: :new: | **app/client/src/constants/CommentConstants.tsx** | **100** | **100** | **100** | **100**
>  :green_circle: | app/client/src/constants/DefaultTheme.tsx | 73.12 **(1.38)** | 80.56 **(0)** | 42.86 **(3.57)** | 75.9 **(0.29)**
>  :sparkles: :new: | **app/client/src/constants/WebsocketConstants.tsx** | **57.14** | **100** | **0** | **100**
>  :green_circle: | app/client/src/constants/messages.ts | 51.48 **(0.28)** | 100 **(0)** | 2.38 **(0.57)** | 72.08 **(0.23)**
>  :sparkles: :new: | **app/client/src/mockResponses/CommentApiMockResponse.tsx** | **100** | **100** | **100** | **100**
>  :green_circle: | app/client/src/reducers/entityReducers/pageListReducer.tsx | 19.44 **(5.55)** | 27.78 **(27.78)** | 15.38 **(15.38)** | 21.88 **(6.25)**
>  :green_circle: | app/client/src/reducers/uiReducers/editorReducer.tsx | 12.12 **(3.03)** | 0 **(0)** | 9.09 **(9.09)** | 12.12 **(3.03)**
>  :sparkles: :new: | **app/client/src/reducers/uiReducers/websocketReducer.ts** | **83.33** | **100** | **0** | **83.33**
>  :sparkles: :new: | **app/client/src/reducers/uiReducers/commentsReducer/commentsReducer.ts** | **70.97** | **0** | **71.43** | **73.33**
>  :sparkles: :new: | **app/client/src/reducers/uiReducers/commentsReducer/handleAddCommentToThreadSuccess.ts** | **100** | **100** | **100** | **100**
>  :sparkles: :new: | **app/client/src/reducers/uiReducers/commentsReducer/handleCreateNewCommentThreadSuccess.ts** | **100** | **100** | **100** | **100**
>  :sparkles: :new: | **app/client/src/reducers/uiReducers/commentsReducer/handleFetchApplicationCommentsSuccess.ts** | **100** | **100** | **100** | **100**
>  :sparkles: :new: | **app/client/src/reducers/uiReducers/commentsReducer/handleNewCommentThreadEvent.ts** | **37.5** | **100** | **0** | **37.5**
>  :sparkles: :new: | **app/client/src/reducers/uiReducers/commentsReducer/handleUpdateCommentThreadEvent.ts** | **42.86** | **100** | **0** | **42.86**
>  :sparkles: :new: | **app/client/src/reducers/uiReducers/commentsReducer/handleUpdateCommentThreadSuccess.ts** | **100** | **100** | **100** | **100**
>  :sparkles: :new: | **app/client/src/reducers/uiReducers/commentsReducer/testFixtures.ts** | **100** | **100** | **100** | **100**
>  :green_circle: | app/client/src/sagas/ApplicationSagas.tsx | 24.4 **(0.3)** | 7.84 **(0)** | 15.79 **(0)** | 27.27 **(0.32)**
>  :green_circle: | app/client/src/sagas/InitSagas.ts | 26.67 **(0.23)** | 3.57 **(0)** | 25 **(0)** | 28.92 **(0.17)**
>  :sparkles: :new: | **app/client/src/sagas/WebsocketSagas.ts** | **20.9** | **9.52** | **6.67** | **21.82**
>  :green_circle: | app/client/src/sagas/index.tsx | 92.5 **(3.03)** | 50 **(0)** | 100 **(0)** | 94.87 **(2.98)**
>  :red_circle: | app/client/src/sagas/userSagas.tsx | 17.37 **(-0.1)** | 2.33 **(-0.37)** | 8.33 **(0)** | 20.57 **(-0.14)**
>  :sparkles: :new: | **app/client/src/sagas/CommentSagas/handleCommentEvents.ts** | **25** | **0** | **0** | **25**
>  :sparkles: :new: | **app/client/src/sagas/CommentSagas/index.ts** | **72.37** | **55** | **75** | **73.33**
>  :green_circle: | app/client/src/selectors/appViewSelectors.tsx | 68.75 **(12.5)** | 0 **(0)** | 28.57 **(28.57)** | 66.67 **(6.67)**
>  :green_circle: | app/client/src/selectors/editorSelectors.tsx | 33.91 **(1.74)** | 0 **(0)** | 4.76 **(4.76)** | 36.45 **(1.87)**
>  :green_circle: | app/client/src/utils/AppsmithUtils.tsx | 40.99 **(1.86)** | 12.2 **(1.22)** | 30.56 **(2.78)** | 39.87 **(1.96)**
>  :sparkles: :new: | **app/client/src/utils/hooks/useIsScrolledToBottom.tsx** | **100** | **72.73** | **100** | **100**
>  :red_circle: | app/client/src/widgets/BaseWidget.tsx | 43.75 **(-0.37)** | 0 **(0)** | 3.45 **(-0.25)** | 45.37 **(-0.55)**
>  :green_circle: | app/client/test/customQueries.js | 95 **(5)** | 100 **(0)** | 66.67 **(33.34)** | 94.74 **(5.27)**

**Comment by jrexerpted** [comment]
folks", "y'all" or "people"). Always be respectful to each other, no matter the communication channel. If you feel like there was a violation of the Code of Conduct, please report it anytime.

We are a community and we're here for you.

Find our code on GitHub
Find ways to contribute

Find us on Twitter

**Comment by jrexerpted** [comment]
folks", "y'all" or "people"). Always be respectful to each other, no matter the communication channel. If you feel like there was a violation of the Code of Conduct, please report it anytime.

We are a community and we're here for you.

Find our code on GitHub
Find ways to contribute

Find us on Twitter

**Comment by jrexerpted** [comment]
folks", "y'all" or "people"). Always be respectful to each other, no matter the communication channel. If you feel like there was a violation of the Code of Conduct, please report it anytime.

We are a community and we're here for you.

Find our code on GitHub
Find ways to contribute

- [ ] - - [ ] Find us on Twitter@ for it for me!  I'm sorry I didn't get back in time to go over to the neighborhood and puke like a choked up lunitic!

File: app/client/src/components/ads/Comments/utils.ts [hunk 1]
```diff
@@ -0,0 +1,44 @@
...
+        ...res.commentThreadsMap,
+        [curr.id]: curr,
+      },
+      refCommentThreads: {
+        ...(res.refCommentThreads ? res.refCommentThreads : {}),
+        [curr.refId]: [
+          ...(res.refCommentThreads && res.refCommentThreads[curr.refId]
+            ? res.refCommentThreads[curr.refId]
+            : []),
+          curr.id,
```
Comment (hetunandu) [inline app/client/src/components/ads/Comments/utils.ts:17]: No need to spread in reduce functions

File: app/client/src/utils/hooks/useIsScrolledToBottom.tsx [hunk 2]
```diff
@@ -0,0 +1,28 @@
+import { useEffect, useState } from "react";
```
Comment (hetunandu) [inline app/client/src/utils/hooks/useIsScrolledToBottom.tsx:1]: I think this us declared in a function as well

File: app/client/src/actions/commentActions.ts [hunk 3]
```diff
@@ -0,0 +1,111 @@
+import { ReduxActionTypes } from "constants/ReduxActionConstants";
+import { COMMENT_EVENTS_CHANNEL } from "constants/CommentConstants";
+import { CommentThread } from "components/ads/Comments/CommentsInterfaces";
+
+export const setCommentThreadsRequest = () => ({
+  type: ReduxActionTypes.SET_COMMENT_THREADS_REQUEST,
+});
+
+export const setCommentThreadsSuccess = (payload: any) => ({
```
[the remark this answers is not in the document]
Comment (rishabhsaxena) [reply app/client/src/actions/commentActions.ts:19]: This ones for dev, would remove it.

File: app/client/src/utils/hooks/useIsScrolledToBottom.tsx (hunk 2 above)
Comment (rishabhsaxena) [reply app/client/src/utils/hooks/useIsScrolledToBottom.tsx:1]: Yeah it was within the same file, had refactored it to a separate file

**Review by hetunandu — COMMENTED** [review]
Looks good to me. Let's not merge it in this release though

**Review by sharat87 — APPROVED** [review]

**Pull request merged**